com.unity.perception/Runtime/Randomization/Scenarios/ScenarioBase.cs

```diff
@@ -1,29 +1,53 @@
 using System;
+using System.Collections.Generic;
+using Unity.Simulation;
 using UnityEngine;
+using UnityEngine.Experimental.Perception.Randomization.Randomizers;
+using UnityEngine.Experimental.Perception.Randomization.Samplers;
 using UnityEngine.Perception.GroundTruth;
-using UnityEngine.Experimental.Perception.Randomization.Configuration;
-using UnityEngine.Experimental.Perception.Randomization.Parameters;
 
 namespace UnityEngine.Experimental.Perception.Randomization.Scenarios
 {
     /// <summary>
-    /// The base class of all scenario classes
+    /// Derive ScenarioBase to implement a custom scenario
     /// </summary>
+    [DefaultExecutionOrder(-1)]
     public abstract class ScenarioBase : MonoBehaviour
     {
         static ScenarioBase s_ActiveScenario;
+
         bool m_SkipFrame = true;
         bool m_FirstScenarioFrame = true;
+        bool m_WaitingForFinalUploads;
+        RandomizerTagManager m_TagManager = new RandomizerTagManager();
+
+        IEnumerable<Randomizer> activeRandomizers
+        {
+            get
+            {
+                foreach (var randomizer in m_Randomizers)
+                    if (randomizer.enabled)
+                        yield return randomizer;
+            }
+        }
+
+        // ReSharper disable once InconsistentNaming
+        [SerializeReference] internal List<Randomizer> m_Randomizers = new List<Randomizer>();
```

Comment:
There are a few places where we scroll through this list looking for randomizers of a certain type, so why not store randomizers in a Dictionary? Also, can we have more than one randomizer of a given type? because get randomizer only returns the first one of a type that is found.

Reply:
First, dictionaries are not a serializable type in Unity so I have to use a list to keep track of the randomizers.

Second, there is only one randomizer of each type to replicate the division of responsibilities methodology found in Unity's official ECS implementation. If there is a need to add two randomizers of the same type, this is often an indicator that the randomizer itself is not properly encapsulating all the responsibilities it should.